File: app/src/main/java/com/example/josycom/flowoverstack/adapters/SearchAdapter.java
package com.example.josycom.flowoverstack.adapters;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.Glide;
import com.example.josycom.flowoverstack.R;
import com.example.josycom.flowoverstack.model.Owner;
import com.example.josycom.flowoverstack.model.Question;
import com.example.josycom.flowoverstack.util.DateUtil;

import org.jsoup.Jsoup;

import java.util.List;

public class SearchAdapter extends RecyclerView.Adapter<SearchAdapter.SearchViewHolder> {

    private static View.OnClickListener mOnClickListener;
    private List<Question> mQuestions;

    @NonNull
    @Override
    public SearchAdapter.SearchViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.question_item, parent, false);
        return new SearchViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull SearchAdapter.SearchViewHolder holder, int position) {
        if (mQuestions != null) {
            Question currentQuestion = mQuestions.get(position);
            holder.bind(currentQuestion);
        }
    }

    public void setOnClickListener(View.OnClickListener mOnClickListener) {
        SearchAdapter.mOnClickListener = mOnClickListener;
    }

    @Override
    public int getItemCount() {
        if (mQuestions != null) {
            return mQuestions.size();
        } else {
            return 0;
        }
    }

    public void setQuestions(List<Question> questions) {
        mQuestions = questions;
        notifyDataSetChanged();
    }

    static class SearchViewHolder extends RecyclerView.ViewHolder {

        ImageView mAvatarView;
        TextView mTitleQuestionText, mViewCounterText, mDateText, mNameText, mAnswersCountText, mVotesCountText, mTagsText;

        SearchViewHolder(@NonNull View itemView) {
            super(itemView);

            mAvatarView = itemView.findViewById(R.id.iv_avatar_item);
            mTitleQuestionText = itemView.findViewById(R.id.tv_question_item);
            mViewCounterText = itemView.findViewById(R.id.tv_counter_item);
            mDateText = itemView.findViewById(R.id.tv_date_item);
            mNameText = itemView.findViewById(R.id.tv_name_item);
            mAnswersCountText = itemView.findViewById(R.id.tv_answers_count_item);
            mVotesCountText = itemView.findViewById(R.id.tv_votes_count_item);
            mTagsText = itemView.findViewById(R.id.tv_tags_list_item);

            itemView.setTag(this);
            itemView.setOnClickListener(mOnClickListener);
        }

        void bind(Question question) {
            if (question != null) {
                Owner owner = question.getOwner();
                String profileImage = owner.getProfileImage();
                List<String> tagList = question.getTags();
                Glide.with(itemView.getContext())
                        .load(profileImage)
                        .placeholder(R.drawable.loading)
                        .into(mAvatarView);
                mTitleQuestionText.setText(Jsoup.parse(question.getTitle()).text());
                mViewCounterText.setText(String.valueOf(question.getViewCount()));
                mDateText.setText(DateUtil.toNormalDate(question.getCreationDate()));
                mNameText.setText(owner.getDisplayName());
                mAnswersCountText.setText(String.valueOf(question.getAnswerCount()));
                mVotesCountText.setText(String.valueOf(question.getScore()));
                mTagsText.setText(updateTagsTextView(tagList));
            }
        }

        private String updateTagsTextView(List<String> tagList) {
            StringBuilder builder = new StringBuilder();
            builder.append("Tags: ");
            for (int i = 0; i < tagList.size(); i++) {
                builder.append(tagList.get(i));
                if (i != tagList.size() - 1) {
                    builder.append(", ");
                }
            }
            return builder.toString();
        }
    }
}

File: app/src/main/java/com/josycom/mayorjay/flowoverstack/view/search/SearchAdapter.kt
package com.josycom.mayorjay.flowoverstack.view.search

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.core.view.isInvisible
import androidx.core.view.isVisible
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.josycom.mayorjay.flowoverstack.R
import com.josycom.mayorjay.flowoverstack.databinding.QuestionItemBinding
import com.josycom.mayorjay.flowoverstack.data.model.Question
import com.josycom.mayorjay.flowoverstack.view.search.SearchAdapter.SearchViewHolder
import com.josycom.mayorjay.flowoverstack.util.AppUtils
import org.jsoup.Jsoup

class SearchAdapter : RecyclerView.Adapter<SearchViewHolder>() {

    private var questions: List<Question>? = null

    companion object {
        private var clickListener: View.OnClickListener? = null
        private var shareClickListener: View.OnClickListener? = null
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): SearchViewHolder {
        val questionItemBinding = QuestionItemBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        return SearchViewHolder(questionItemBinding)
    }

    override fun onBindViewHolder(holder: SearchViewHolder, position: Int) {
        if (questions != null) {
            val currentQuestion = questions?.get(position)
            holder.bind(currentQuestion)
        }
    }

    fun setOnClickListener(onClickListener: View.OnClickListener?, shareOnClickListener: View.OnClickListener?) {
        clickListener = onClickListener
        shareClickListener = shareOnClickListener
    }

    override fun getItemCount(): Int {
        return if (questions != null) questions?.size ?: 0 else 0
    }

    fun setQuestions(questions: List<Question>?) {
        this.questions = questions
        notifyDataSetChanged()
    }

    inner class SearchViewHolder(private val binding: QuestionItemBinding) : RecyclerView.ViewHolder(binding.root) {

        init {
            binding.root.tag = this
            binding.root.setOnClickListener(clickListener)
            binding.ivShare.setOnClickListener(shareClickListener)
        }

        fun bind(question: Question?) {
            if (question != null) {
                binding.ivShare.tag = question
                val owner = question.owner
                val profileImage = owner.profileImage
                val tagList = question.tags
                Glide.with(binding.root.context)
                    .load(profileImage)
                    .placeholder(R.drawable.loading)
                    .into(binding.ivAvatarItem)
                binding.tvQuestionItem.text = Jsoup.parse(question.title).text()
                binding.tvViewsCountItem.text = question.viewCount.toString()
                binding.tvDateItem.text = AppUtils.toNormalDate(question.creationDate.toLong())
                if (question.isAnswered) {
                    binding.answered.isVisible = true
                } else {
                    binding.answered.isInvisible = true
                }
                val answers = question.answerCount
                val resources = binding.root.context.resources
                val answerCount = resources.getQuantityString(R.plurals.answers, answers, answers)
                binding.tvAnswersCountItem.text = answerCount
                if (question.score <= 0) {
                    binding.tvVotesCountItem.text = question.score.toString()
                } else {
                    binding.tvVotesCountItem.text =
                        binding.root.context.getString(R.string.plus_score, question.score)
                }
                binding.tvTagsListItem.text = AppUtils.getFormattedTags(tagList)
            }
        }
    }
}